File: src/main/java/cs/android/view/list/CSListController.java
package cs.android.view.list;

import android.content.res.Configuration;
import android.util.SparseBooleanArray;
import android.view.View;
import android.view.View.OnLayoutChangeListener;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.ListView;

import java.util.List;

import cs.android.lang.CSWork;
import cs.android.viewbase.CSView;
import cs.android.viewbase.CSViewController;
import cs.java.callback.CSReturnWith;
import cs.java.callback.CSRunWithWith;
import cs.java.collections.CSList;
import cs.java.event.CSEvent;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

import static android.os.SystemClock.uptimeMillis;
import static android.view.MotionEvent.ACTION_CANCEL;
import static android.view.MotionEvent.obtain;
import static cs.java.lang.CSLang.NO;
import static cs.java.lang.CSLang.YES;
import static cs.java.lang.CSLang.event;
import static cs.java.lang.CSLang.fire;
import static cs.java.lang.CSLang.is;
import static cs.java.lang.CSLang.iterate;
import static cs.java.lang.CSLang.list;
import static cs.java.lang.CSLang.no;

public class CSListController<RowType, T extends AbsListView> extends CSViewController<T> {
    public final CSEvent<List<RowType>> onLoad = event();
    private final CSList<RowType> _dataList = list();
    private CSReturnWith<CSIRowView<RowType>, Integer> _createView;
    private BaseAdapter _listAdapter = new CSListAdapter(this);
    private int _savedSelectionIndex;
    private int _firstVisiblePosition;
    private int _itemsTypesCount = 1;
    private CSView _emptyView;
    private boolean _firstLoad;
    private SparseBooleanArray _savedCheckedItems;
    private CSRunWithWith<Integer, CSIRowView<RowType>> _onItemClick;
    private CSReturnWith<Integer, Integer> _positionViewType;
    private CSRunWithWith<Integer, CSIRowView<RowType>> _onItemLongClick;
    private CSReturnWith<Boolean, Integer> _isEnabled;
    private CSWork _autoReload;

    public CSListController(CSViewController parent, int listViewId, CSReturnWith<CSIRowView<RowType>, Integer> createView) {
        super(parent, listViewId);
        _createView = createView;
    }

    public CSListController(CSViewController parent, int listViewId) {
        super(parent, listViewId);
    }

    public CSListController<RowType, T> onCreateView(CSReturnWith<CSIRowView<RowType>, Integer> createView) {
        _createView = createView;
        return this;
    }

    public BaseAdapter adapter() {
        return _listAdapter;
    }

    public void clear() {
        _dataList.clear();
        reloadData();
    }

    public CSList<RowType> data() {
        return _dataList;
    }

    public CSView emptyView() {
        return _emptyView;
    }

    public CSList<RowType> getCheckedRows() {
        final CSList<RowType> checkedRows = list();
        SparseBooleanArray positions = asView().getCheckedItemPositions();
        if (positions != null) for (int i = 0; i < positions.size(); i++)
            if (positions.valueAt(i)) {
                RowType checkedRow = data().at(positions.keyAt(i));
                if (is(checkedRow)) checkedRows.add(checkedRow);
            }
        return checkedRows;
    }

    int getItemTypesCount() {
        return _itemsTypesCount;
    }

    @SuppressWarnings({"unchecked"})
    public View getRowView(int position, View view) {
        CSIRowView<RowType> rowView;
        if (no(view)) {
            rowView = createView(getItemViewType(position));
            view = rowView.asView();
            view.setTag(rowView);
        } else rowView = asRowView(view);
        rowView.row(_dataList.get(position));
        return view;
    }

    @SuppressWarnings({"unchecked"})
    private CSIRowView<RowType> asRowView(View view) {
        return (CSIRowView<RowType>) view.getTag();
    }

    public CSListController<RowType, T> loadAdd(List<RowType> list) {
        _firstLoad = YES;
        _dataList.addAll(list);
        reloadData();
        fire(onLoad, list);
        return this;
    }

    public void prependData(RowType status) {
        _dataList.add(0, status);
        reloadData();
    }

    public CSListController<RowType, T> reload(List<RowType> list) {
//        saveState();
        _dataList.clear();
        loadAdd(list);
//        restoreState();
        return this;
    }

    public void reloadData() {
        _listAdapter.notifyDataSetChanged();
        updateEmptyView();
    }

    public void restoreSelectionAndScrollState() {
        if (asView() instanceof ListView)
            asListView().setSelectionFromTop(_firstVisiblePosition, 0);
        if (_savedSelectionIndex > -1) asAbsListView().setSelection(_savedSelectionIndex);
        if (is(_savedCheckedItems)) for (int i : iterate(_savedCheckedItems.size()))
            if (_savedCheckedItems.valueAt(i)) asAbsListView()
                    .setItemChecked(i, _savedCheckedItems.valueAt(i));
    }

    public void saveSelectionAndScrollState() {
        if (asView() instanceof ListView)
            _firstVisiblePosition = asListView().getFirstVisiblePosition();
        _savedSelectionIndex = asAbsListView().getSelectedItemPosition();
        _savedCheckedItems = asAbsListView().getCheckedItemPositions();
    }

    public void scrollToTop() {
        asAbsListView().setSelection(0);
        asAbsListView().dispatchTouchEvent(obtain(uptimeMillis(), uptimeMillis(), ACTION_CANCEL, 0, 0, 0));
    }

    public CSView emptyView(int id) {
        return emptyView(new CSView<>(parent(), id));
    }

    public CSView emptyView(CSView view) {
        _emptyView = view;
        if (is(_emptyView)) _emptyView.hide();
        return _emptyView;
    }

    public CSListController<RowType, T> setListAdapter(BaseAdapter listAdapter) {
        _listAdapter = listAdapter;
        return this;
    }

    private void updateEmptyView() {
        if (!_firstLoad) return;
        if (no(_emptyView)) return;
        if (_dataList.isEmpty())
            _emptyView.show();
        else _emptyView.hide();
    }

    protected CSIRowView<RowType> createView(int viewType) {
        return _createView.invoke(viewType);
    }

    protected final int getItemViewType(int position) {
        return is(_positionViewType) ? _positionViewType.invoke(position) : position;
    }

    protected void onCreate() {
        super.onCreate();
        asAdapterView().setAdapter(_listAdapter);
        asAbsListView().setFastScrollEnabled(YES);
        asAbsListView().setOnItemClickListener((parent, view, position, id) -> {
            if (is(_onItemClick)) _onItemClick.run(position, asRowView(view));
        });
        asAbsListView().setOnItemLongClickListener((parent, view, position, id) -> {
            if (is(_onItemLongClick)) _onItemLongClick.run(position, asRowView(view));
            return true;
        });
    }

    int getItemsCount() {
        return _dataList.size();
    }

    public CSListController<RowType, T> viewTypes(int itemsTypesCount) {
        _itemsTypesCount = itemsTypesCount;
        return this;
    }

    public CSListController<RowType, T> onItemClick(CSRunWithWith<Integer, CSIRowView<RowType>> onItemClick) {
        _onItemClick = onItemClick;
        return this;
    }

    public CSListController<RowType, T> onItemLongClick(CSRunWithWith<Integer, CSIRowView<RowType>> onItemLongClick) {
        _onItemLongClick = onItemLongClick;
        return this;
    }

    public CSListController<RowType, T> positionViewType(CSReturnWith<Integer, Integer> getTtemViewType) {
        _positionViewType = getTtemViewType;
        return this;
    }

    public RowType dataAt(Integer position) {
        return this.data().at(position);
    }

    public CSListController<RowType, T> isEnabled(CSReturnWith<Boolean, Integer> isEnabled) {
        _isEnabled = isEnabled;
        return this;
    }

    public void checkAll() {
        for (Integer index : iterate(asAbsListView().getCount()))
            asAbsListView().setItemChecked(index, YES);
    }

    public void unCheckAll() {
        SparseBooleanArray positions = asView().getCheckedItemPositions();
        if (positions != null) for (int i = 0; i < positions.size(); i++) {
            boolean checked = positions.valueAt(i);
            if (checked) {
                int index = positions.keyAt(i);
                asAbsListView().setItemChecked(index, NO);
            }
        }
    }

    public boolean isEnabled(int position) {
        if (is(_isEnabled)) return _isEnabled.invoke(position);
        return true;
    }

    public CSListController setEmptyText(int emptyView, String message) {
        emptyView(emptyView).text(message);
        return this;
    }

    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        saveSelectionAndScrollState();
        asAdapterView().setAdapter(_listAdapter);
        restoreSelectionAndScrollState();
    }


    public CSListController<RowType, T> onLoadDone(Function1<CSListController<RowType, T>, Unit> function) {
        final OnLayoutChangeListener listener = new OnLayoutChangeListener() {
            public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop, int oldRight, int oldBottom) {
                function.invoke(CSListController.this);
                asView().removeOnLayoutChangeListener(this);
            }
        };
        if (is(function)) asView().addOnLayoutChangeListener(listener);
        else asView().removeOnLayoutChangeListener(listener);
        return this;
    }
}

File: library/renetik-android/src/main/java/renetik/android/view/list/CSListController.kt
package renetik.android.view.list

import android.content.res.Configuration
import android.os.SystemClock.uptimeMillis
import android.util.SparseBooleanArray
import android.view.MotionEvent.ACTION_CANCEL
import android.view.MotionEvent.obtain
import android.view.View
import android.widget.AbsListView
import android.widget.BaseAdapter
import android.widget.ListView
import renetik.android.extensions.findView
import renetik.android.extensions.view.hide
import renetik.android.extensions.view.show
import renetik.android.view.base.CSViewController
import renetik.java.collections.CSList
import renetik.java.collections.list
import renetik.java.event.event

open class CSListController<RowType : Any, ViewType : AbsListView> : CSViewController<ViewType> {

    val onLoad = event<List<RowType>>()
    val data = list<RowType>()
    var viewTypesCount = 1
    var listAdapter: BaseAdapter = CSListAdapter(this)
    private var firstLoad = false
    private lateinit var createView: (Int) -> CSRowView<RowType>
    private var firstVisiblePosition: Int = 0
    private var emptyView: View? = null
        set(value) {
            field = value?.hide()
        }
    private var onItemClick: ((Int, CSRowView<RowType>) -> Unit)? = null
    private var onPositionViewType: ((Int) -> Int)? = null
    private var onItemLongClick: ((Int, CSRowView<RowType>) -> Unit)? = null
    private var onIsEnabled: ((Int) -> Boolean)? = null
    private var savedSelectionIndex: Int = 0
    private var savedCheckedItems: SparseBooleanArray? = null

    val checkedRows: CSList<RowType>
        get() {
            val checkedRows = list<RowType>()
            val positions = view.checkedItemPositions
            if (positions != null)
                for (i in 0 until positions.size())
                    if (positions.valueAt(i))
                        data.at(positions.keyAt(i))?.let { checkedRow -> checkedRows.add(checkedRow) }
            return checkedRows
        }

    constructor(parent: CSViewController<*>, view: ViewType, createView: (Int) -> CSRowView<RowType>)
            : super(parent, view) {
        this.createView = createView
    }

    constructor(parent: CSViewController<*>, listViewId: Int) : super(parent, listViewId) {}

    fun onCreateView(function: (Int) -> CSRowView<RowType>) = apply { createView = function }

    fun clear() = apply {
        data.clear()
        reloadData()
    }

    fun getRowView(position: Int, view: View?): View {
        var rowView: CSRowView<RowType>
        if (view == null) rowView = createView.invoke(getItemViewType(position))
        else rowView = asRowView(view)
        rowView.data(data.get(position))
        return rowView.view
    }

    private fun asRowView(view: View) = view.tag as CSRowView<RowType>

    fun loadAdd(list: List<RowType>) = apply {
        firstLoad = true
        data.addAll(list)
        reloadData()
        onLoad.fire(list)
        return this
    }

    fun prependData(status: RowType) = apply {
        data.add(0, status)
        reloadData()
    }

    fun reload(list: List<RowType>) = apply {
        data.clear()
        loadAdd(list)
    }

    fun reloadData() {
        listAdapter.notifyDataSetChanged()
        updateEmptyViewVisibility()
    }

    fun restoreSelectionAndScrollState() {
        (view as? ListView)?.setSelectionFromTop(firstVisiblePosition, 0)
        if (savedSelectionIndex > -1) view.setSelection(savedSelectionIndex)
        savedCheckedItems?.let { item ->
            for (i in 0 until item.size())
                if (item.valueAt(i)) view.setItemChecked(i, item.valueAt(i))
        }
    }

    fun saveSelectionAndScrollState() {
        (view as? ListView)?.let { firstVisiblePosition = it.firstVisiblePosition }
        savedSelectionIndex = view.selectedItemPosition
        savedCheckedItems = view.checkedItemPositions
    }

    fun scrollToTop() {
        view.setSelection(0)
        view.dispatchTouchEvent(obtain(uptimeMillis(), uptimeMillis(), ACTION_CANCEL, 0f, 0f, 0))
    }

    fun getItemViewType(position: Int) = onPositionViewType?.invoke(position) ?: position

    override fun onCreate() {
        super.onCreate()
        view.setAdapter(listAdapter)
        view.isFastScrollEnabled = true
        view.setOnItemClickListener { _, view, position, _ ->
            onItemClick?.invoke(position, asRowView(view))
        }
        view.setOnItemLongClickListener { _, view, position, _ ->
            onItemLongClick?.invoke(position, asRowView(view))
            true
        }
    }

    fun onItemClick(function: (Int, CSRowView<RowType>) -> Unit) = apply { onItemClick = function }

    fun onItemLongClick(function: (Int, CSRowView<RowType>) -> Unit) = apply { onItemLongClick = function }

    fun onPositionViewType(function: (Int) -> Int) = apply { onPositionViewType = function }

    fun dataAt(position: Int) = data.at(position)

    fun onIsEnabled(function: (Int) -> Boolean) = apply { onIsEnabled = function }

    fun isEnabled(position: Int) = onIsEnabled?.invoke(position) ?: true

    fun emptyView(id: Int) = apply { emptyView = parentController?.findView(id) }

    fun checkAll() = apply { for (index in 0 until view.count) view.setItemChecked(index, true) }

    fun unCheckAll() = apply {
        val positions = view.checkedItemPositions
        if (positions != null)
            for (i in 0 until positions.size()) {
                val checked = positions.valueAt(i)
                if (checked) {
                    val index = positions.keyAt(i)
                    view.setItemChecked(index, false)
                }
            }
    }

    private fun updateEmptyViewVisibility() {
        if (!firstLoad) return
        emptyView?.let {
            if (data.isEmpty()) it.show()
            else it.hide()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        saveSelectionAndScrollState()
        view.setAdapter(listAdapter)
        restoreSelectionAndScrollState()
    }

}
